### DISCUSSION TITLE
how about let mutation operation has transaction ability automatically?

### DISCUSSION THREAD

**Discussion by bonafideyan** [99designs/gqlgen#2310] [Q&A] [answered]
### What happened?
nothing
### What did you expect?
I can omit lots lines for transaction, such as db.rollback, db.commit

**Comment by giautm** [comment, marked as answer]
Hi @bonafideyan, because it depends on the Database driver / ORM that you used in your project. So, I think GQLGen can't do it automatically. However, you can apply the flowing pattern:

1. Open Transaction use graphql.HandlerExtension, and attach it to the Go's context object.
2. In your mutations, you can get the Tx object from the context and do the updates. If something go wrong, just return the error and the extension in step 1) will call rollback automatic

We (Ent/EntGQL) applied this pattern in our implementation. So, you can check out [the example here for detail](https://github.com/ent/contrib/blob/master/entgql/transaction.go) and [here](https://github.com/ent/contrib/blob/master/entgql/internal/todo/server/server.go#L57). This also can apply to the regular SQL driver:

**Reply by bonafideyan** [reply]
Thank you. That is a perfect solution.
I have read it, but don't know why: 
     at line 94 of file transaction.go,  panic(r) after rollback, does it terminate the process? I usually write it as rollback and log it, can I remove the panic line?

**Reply by giautm** [reply]
Yes, you can remove the panic and log the error.
